src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -12912,12 +12912,35 @@ private ExpressionSyntax ParseObjectOrCollectionInitializerMember(ref bool isObj
         private PostSkipAction SkipBadInitializerListTokens<T>(ref SyntaxToken startToken, SeparatedSyntaxListBuilder<T> list, SyntaxKind expected)
             where T : CSharpSyntaxNode
         {
+            bool openBraceIsMissing = startToken.IsMissing;
             return this.SkipBadSeparatedListTokensWithExpectedKind(ref startToken, list,
                 p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleExpression(),
-                p => p.CurrentToken.Kind == SyntaxKind.CloseBraceToken || p.IsTerminator(),
+                p => p.IsEndOfInitializer(openBraceIsMissing),
                 expected);
         }
 
+        private bool IsEndOfInitializer(bool openBraceIsMissing)
+        {
+            if (!openBraceIsMissing && CurrentToken.Kind == SyntaxKind.SemicolonToken)
+            {
+                // When encountering a semicolon in an itializer body, the user may have mistakenly intended it as a separator.
+
+                // Seek ahead and check if the semicolon is being followed by:
+                // 1. a closing brace - the semicolon is not the end, the closing brace is the end
+                // 2. another variable intializer - not the end yet
+                var resetPoint = GetResetPoint();
+                EatToken();
+                bool initializerContinues = CurrentToken.Kind == SyntaxKind.CloseBraceToken || IsPossibleVariableInitializer();
```

Comment:
Note: this looks like it woudl barf on `new Dict { [0] = "a"; [1] = "B" }`.   i'd prefer us be more intentional in what we're looking for as the potential start of a next initializer.

Reply:
added tests that show it doesnt barf now